server/s3_list_object_versions_action.cc

```diff
@@ -843,13 +839,10 @@ std::string& S3ListObjectVersionsAction::get_response_xml() {
 // Delimiter, KeyMarker, Prefix, NextKeyMarker, Key.
 std::string S3ListObjectVersionsAction::get_encoded_key_value(
     const std::string& key_value) {
-  std::string encoded_key_value;
-  if (encoding_type == "url") {
-    char* encoded_str = evhttp_uriencode(key_value.c_str(), -1, 0);
-    encoded_key_value = encoded_str;
-    free(encoded_str);
-  } else {
-    encoded_key_value = key_value;
+  if (encoding_type != "url") {
+    return key_value;
   }
-  return encoded_key_value;
+  const std::unique_ptr<char, void (*)(void*)> output(
+      evhttp_uriencode(key_value.c_str(), -1, 0), &free);
+  return output.get();
```

Comment:
If `evhttp_uriencode` fails for whatever reason returns a NULL pointer, and you can't construct a std::string from NULL. I think this should be checking for NULL even though other code doesn't. 

```suggestion
  std::string encoded_key_value;
  const std::unique_ptr<char, void (*)(void*)> encoded_str(
      evhttp_uriencode(key_value.c_str(), -1, 0), &free);
  if (encoded_str) {
    encoded_key_value = encoded_str;
  }
  return encoded_key_value;
```

Reply:
Updated.